pravega/flink-connectors
# [issue-31] Implement connector metrics

Signed-off-by: Vijay Srinivasaraghavan <vijayaraghavan.srinivasaraghavan@emc.com>

  * Added support to draw reader group metrics for FlinkPravegaReader
  * Following metrics are enabled
    - reader group name
    - scope
    - stream names
    - online readers
    - segment positions
    - unread bytes

**Change log description**
Fixes #31

**Purpose of the change**
To publish reader group metrics

**What the code does**
Publishes the runtime stats of the reader group as Flink metrics

**How to verify it**
- On a running Flink cluster alongside with Pravega, run a job that uses `FlinkPravegaReader`  (e.g., anomaly-samples application) and verify if the metrics are visible from UI as well as REST API
```
curl -i -s -f localhost:1028/jobs/2c248d6fe8ed518e959989889a3c5ced/vertices/bc764cd8ddf7a0cff126f51c16239658/metrics?get=0.Source__AnomalyEventReader.PravegaReader.readerGroup.readerGroupName
[{"id":"0.Source__AnomalyEventReader.PravegaReader.readerGroup.readerGroupName","value":"flinkmrdf01aislmcira301jg"}]

curl -i -s -f localhost:1028/jobs/2c248d6fe8ed518e959989889a3c5ced/vertices/bc764cd8ddf7a0cff126f51c16239658/metrics?get=0.Source__AnomalyEventReader.PravegaReader.readerGroup.scope
[{"id":"0.Source__AnomalyEventReader.PravegaReader.readerGroup.scope","value":"anomaly"}]

curl -i -s -f localhost:1028/jobs/2c248d6fe8ed518e959989889a3c5ced/vertices/bc764cd8ddf7a0cff126f51c16239658/metrics?get=0.Source__AnomalyEventReader.PravegaReader.readerGroup.streams
[{"id":"0.Source__AnomalyEventReader.PravegaReader.readerGroup.streams","value":"test"}]

curl -i -s -f localhost:1028/jobs/2c248d6fe8ed518e959989889a3c5ced/vertices/bc764cd8ddf7a0cff126f51c16239658/metrics?get=0.Source__AnomalyEventReader.PravegaReader.readerGroup.onlineReaders
[{"id":"0.Source__AnomalyEventReader.PravegaReader.readerGroup.onlineReaders","value":"Source: AnomalyEventReader (1/1)"}]

curl -i -s -f localhost:1028/jobs/2c248d6fe8ed518e959989889a3c5ced/vertices/bc764cd8ddf7a0cff126f51c16239658/metrics?get=0.Source__AnomalyEventReader.PravegaReader.readerGroup.stream.test.segmentPositions
[{"id":"0.Source__AnomalyEventReader.PravegaReader.readerGroup.stream-test.segmentPositions","value":"stream=test, segments={anomaly/test/2.#epoch.0=0 ,anomaly/test/0.#epoch.0=0 ,anomaly/test/1.#epoch.0=0 ,}"}]

curl -i -s -f localhost:1028/jobs/2c248d6fe8ed518e959989889a3c5ced/vertices/bc764cd8ddf7a0cff126f51c16239658/metrics?get=0.Source__AnomalyEventReader.PravegaReader.readerGroup.unreadBytes
[{"id":"0.Source__AnomalyEventReader.PravegaReader.readerGroup.unreadBytes","value":"3197"}]

```
